DataFormats/TrackerRecHit2D/src/VectorHit.cc

```diff
@@ -0,0 +1,413 @@
+  Global3DPoint gp = geomDet->surface().toGlobal(lp);
+  return gp;
+}
+
+GlobalError VectorHit::lowerGlobalPosErr() const {
+  const StackGeomDet* stackDet = dynamic_cast<const StackGeomDet*>(det());
+  const PixelGeomDetUnit* geomDetLower = dynamic_cast<const PixelGeomDetUnit*>(stackDet->lowerDet());
+  return phase2clusterGlobalPosErr(geomDetLower);
+}
+
+GlobalError VectorHit::upperGlobalPosErr() const {
+  const StackGeomDet* stackDet = dynamic_cast<const StackGeomDet*>(det());
+  const PixelGeomDetUnit* geomDetUpper = dynamic_cast<const PixelGeomDetUnit*>(stackDet->upperDet());
+  return phase2clusterGlobalPosErr(geomDetUpper);
+}
+
+GlobalError VectorHit::phase2clusterGlobalPosErr(const PixelGeomDetUnit* geomDet) const {
+  const PixelTopology* topo = &geomDet->specificTopology();
+  float pitchX = topo->pitch().first;
+  float pitchY = topo->pitch().second;
+  LocalError le(pow(pitchX, 2) / 12., 0, pow(pitchY, 2) / 12.);  // e2_xx, e2_xy, e2_yy
+  GlobalError ge(ErrorFrameTransformer().transform(le, geomDet->surface()));
+  return ge;
+}
+
+Global3DVector VectorHit::globalDelta() const {
+  Local3DVector theLocalDelta =
+      LocalVector(theDirection.x() * theDirection.z(), theDirection.y() * theDirection.z(), theDirection.z());
+  Global3DVector g = det()->surface().toGlobal(theLocalDelta);
+  return g;
+}
+
+Global3DVector VectorHit::globalDirection() const { return (det()->surface().toGlobal(localDirection())); }
+
+std::pair<double, double> VectorHit::curvatureORphi(std::string curvORphi) const {
+  double curvature = -999.;
+  double errorCurvature = -999.;
+  double phi = -999.;
+
+  //global pos and errors
+  Global3DPoint gPositionLower = lowerGlobalPos();
+  Global3DPoint gPositionUpper = upperGlobalPos();
+
+  GlobalError gErrorLower = lowerGlobalPosErr();
+  GlobalError gErrorUpper = upperGlobalPosErr();
+
+  //insert lower and upper in the global sor
+  if (gPositionLower.perp() > gPositionUpper.perp()) {
+    gPositionLower = upperGlobalPos();
+    gPositionUpper = lowerGlobalPos();
+    gErrorLower = upperGlobalPosErr();
+    gErrorUpper = lowerGlobalPosErr();
+  }
+
+  double h1 = gPositionLower.x() * gPositionUpper.y() - gPositionUpper.x() * gPositionLower.y();
+
+  //determine sign of curvature
+  AlgebraicVector2 n1;
+  n1[0] = -gPositionLower.y();
+  n1[1] = gPositionLower.x();
+  AlgebraicVector2 n2;
+  n2[0] = gPositionUpper.x() - gPositionLower.x();
+  n2[1] = gPositionUpper.y() - gPositionLower.y();
+
+  double n3 = n1[0] * n2[0] + n1[1] * n2[1];
+  double signCurv = -copysign(1.0, n3);
+  double phi1 = atan2(gPositionUpper.y() - gPositionLower.y(), gPositionUpper.x() - gPositionLower.x());
+
+  if (h1 != 0) {
+    double h2 = 2 * h1;
+    double r12 = pow(gPositionLower.x(), 2) + pow(gPositionLower.y(), 2);
+    double r22 = pow(gPositionUpper.x(), 2) + pow(gPositionUpper.y(), 2);
+    double h3 =
+        (pow(gPositionLower.x(), 2) - 2. * gPositionLower.x() * gPositionUpper.x() + pow(gPositionUpper.x(), 2) +
```

Comment:
why not ```pow(gPositionLower.x() - gPositionUpper.x(), 2)``` ? 
The implementation here is quite verbose with the same method calls reused multiple times.
Having a shorter temp variable names can make this code more readable.

Reply:
I have simplified the code here according to the suggestion, and also replaced all these function calls with temporary variables. Much easier to read indeed.